tools/generate_code_cache.js

```diff
@@ -0,0 +1,124 @@
+'use strict';
+
+// Flags: --expose-internals
+
+// This file generates the code cache for builtin modules and
+// writes them into static char arrays of a C++ file that can be
+// compiled into the binary using the `--code-cache-path` option
+// of `configure`.
+
+const {
+  nativeModuleWrap,
+  builtinSource,
+  cannotUseCache
+} = require('internal/bootstrap/cache');
+
+const vm = require('vm');
+const fs = require('fs');
+
+const resultPath = process.argv[2];
+if (!resultPath) {
+  console.error(`Usage: ${process.argv[0]} ${process.argv[1]}` +
+                'path/to/node_code_cache.cc');
+  process.exit(1);
+}
+
+/**
+ * Format a number of a size in bytes into human-readable strings
+ * @param {number} num
+ * @return {string}
+ */
+function formatSize(num) {
+  if (num < 1024) {
+    return `${(num).toFixed(2)}B`;
+  } else if (num < 1024 ** 2) {
+    return `${(num / 1024).toFixed(2)}KB`;
+  } else if (num < 1024 ** 3) {
+    return `${(num / (1024 ** 2)).toFixed(2)}MB`;
+  } else {
+    return `${(num / (1024 ** 3)).toFixed(2)}GB`;
+  }
+}
+
+/**
+ * Generates the source code of definitions of the char arrays
+ * that contains the code cache and the source code of the
+ * initializers of the code cache.
+ *
+ * @param {string} key ID of the builtin module
+ * @param {Buffer} cache Code cache of the builtin module
+ * @return { definition: string, initializer: string }
+ */
+function getInitalizer(key, cache) {
+  const defName = key.replace(/\//g, '_').replace(/-/g, '_');
+  const definition = `static uint8_t ${defName}_raw[] = {\n` +
+                     `${cache.join(',')}\n};`;
+  const initializer = `
+  v8::Local<v8::ArrayBuffer> ${defName}_ab =
+    v8::ArrayBuffer::New(isolate, ${defName}_raw, ${cache.length});
+  v8::Local<v8::Uint8Array> ${defName}_array =
+    v8::Uint8Array::New(${defName}_ab, 0, ${cache.length});
+  target->Set(context,
+              FIXED_ONE_BYTE_STRING(isolate, "${key}"),
+              ${defName}_array).FromJust();
+  `;
+  return {
+    definition, initializer
+  };
+}
+
+const cacheDefinitions = [];
+const cacheInitializers = [];
+let totalCacheSize = 0;
+
+
+for (const key of Object.keys(builtinSource)) {
+  if (cannotUseCache.includes(key)) continue;
+  const code = nativeModuleWrap(builtinSource[key]);
+
+  // Note that this must corresponds to the code in
+  // NativeModule.prototype.compile
+  const script = new vm.Script(code, {
+    filename: `${key}.js`,
+    produceCachedData: true
```

Comment:
Script#createCachedData has landed so i wonder if maybe there's another way to do this so we can snapshot them after initial evaluation.

Reply:
yeah sorry for using some confusing terms. I mean we should see if we can produce code caches right after the js runs but before we run user code or smth similar.

Reply:
@devsnek Does that make a difference though? The options passed to `ScriptCompiler` would probably still be the same and the script is still unbound.

Reply:
@joyeecheung we would make a code cache from after the code first evaluated, which as i understand it is essentially a "fast-forward" to that state when you use the code cache

Reply:
@devsnek Did you mean creating the code cache out of a `Function`? Otherwise I cannot see a difference if the unbound scripts are produced by passing the same arguments to `CompileUnboundScript`? I have not checked but I'd expect the code cache produced here and the code cache produced in the loaders to be the same?

Reply:
@joyeecheung right now we create the caches before evaluation. i'm saying we create them after evaluation.

Reply:
@devsnek Did you mean we can create the code cache after `unbound_script->BindToCurrentContext()` and `bounded_script->Run()`? I am still not sure what difference that makes and how to get the core to spit the cache out to a file without relying on any environment variable or flags in `loaders.js`...maybe we can leave that for a future PR?

Reply:
Let me clear up some confusion :)

This current PR creates the code cache right after each script compiles. At that time, V8 only compiled the toplevel function. Any inner functions that are called during bootstrapping needs to be compiled as we go. We call this lazy compilation. Lazily compiled functions are not part of the code cache.

Script#createCachedData introduces a way to create the code cache at arbitrary time, as long as you have the v8::UnboundScript object. The benefit is that if you call it after you have executed the script for bootstrapping, lazily compiled functions during that execution are now also included in the code cache. Therefore, when bootstrapping with code cache, you no longer need to compile previously lazily compiled functions.

Neither of the two ways of code caching captures state (to fast-forward to). Please do not confuse code cache with startup snapshot :)